libpoolprotocols/PoolManager.cpp

```diff
@@ -136,28 +149,103 @@ PoolManager::PoolManager(PoolClient * client, Farm &farm, MinerType const & mine
 
 void PoolManager::stop()
 {
-	if (m_running) {
+	if (m_running.load(std::memory_order_relaxed)) {
+		
 		cnote << "Shutting down...";
-		m_running = false;
 
-		if (p_client->isConnected())
-			p_client->disconnect();
+		m_running.store(false, std::memory_order_relaxed);
+		if (p_client->isConnected()) 
+		{ 
+
+			// Disconnection will also trigger
+			// stop for mining
+			p_client->disconnect(); 
+
+		}
+		else {
+
+			if (m_farm.isMining())
+			{
+				cnote << "Shutting down miners...";
+				m_farm.stop();
+			}
 
-		if (m_farm.isMining())
-		{
-			cnote << "Shutting down miners...";
-			m_farm.stop();
 		}
+
 	}
 }
 
 void PoolManager::workLoop()
 {
-	while (m_running)
+	while (m_running.load(std::memory_order_relaxed))
 	{
-		this_thread::sleep_for(chrono::seconds(1));
-		m_hashrateReportingTimePassed++;
+
+		// Take action only if not pending state (connecting/disconnecting)
+		// Otherwise do nothing and wait until connection state is NOT pending
+		if (!p_client->isPendingState()) {
+
+			dev::setThreadName("main");
```

Comment:
You don't want to do this in a loop.

Reply:
What about this?

Reply:
Why not ?
It's conditional and happens quite unfrequently. 
Anyway I can't see clearly any other place where to place it to have the thread properly labelled in output logs.

Reply:
Can't you just do it once in the beginning of the function?